server/src/main/java/io/druid/metadata/SQLMetadataStorageActionHandler.java

```diff
@@ -453,44 +477,66 @@ public Boolean withHandle(Handle handle) throws Exception
+                        "SELECT log_payload_id, log_payload_order, log_payload_part FROM %1$s WHERE %2$s_id = :entryId",
                         logTable, entryTypeName
                     )
                 )
                 .bind("entryId", entryId)
-                .map(ByteArrayMapper.FIRST)
+                .map((index, r, ctx) -> new LogPayloadPart(r.getString(1), r.getInt(2), r.getBytes(3)))
                 .fold(
-                    Lists.<LogType>newLinkedList(),
-                    new Folder3<List<LogType>, byte[]>()
-                    {
-                      @Override
-                      public List<LogType> fold(
-                          List<LogType> list, byte[] bytes, FoldController control, StatementContext ctx
-                      ) throws SQLException
-                      {
-                        try {
-                          list.add(
-                              jsonMapper.<LogType>readValue(
-                                  bytes, logType
-                              )
-                          );
-                          return list;
-                        }
-                        catch (IOException e) {
-                          log.makeAlert(e, "Failed to deserialize log")
-                             .addData("entryId", entryId)
-                             .addData("payload", StringUtils.fromUtf8(bytes))
-                             .emit();
-                          throw new SQLException(e);
-                        }
-                      }
+                    new TreeMap<>(),
+                    (Folder3<Map<String, List<LogPayloadPart>>, LogPayloadPart>)
+                        (accumulator, payloadPart, control, ctx) -> {
+                      accumulator.computeIfAbsent(payloadPart.id, k -> new ArrayList<>())
+                                 .add(payloadPart);
+                      return accumulator;
                     }
-                );
+                )
+                .values()
+                .stream()
+                .map(payloadParts -> {
+                  payloadParts.sort(Comparator.comparing(part -> part.order));
+                  final byte[] payload = new byte[payloadParts.stream().mapToInt(part -> part.part.length).sum()];
+                  int offset = 0;
+                  for (LogPayloadPart payloadPart : payloadParts) {
+                    final byte[] part = payloadPart.part;
+                    System.arraycopy(part, 0, payload, offset, part.length);
+                    offset += part.length;
+                  }
+                  try {
+                    return jsonMapper.<LogType>readValue(payload, logType);
+                  }
+                  catch (IOException e) {
+                    log.makeAlert(e, "Failed to deserialize log")
+                       .addData("entryId", entryId)
+                       .addData("payload", StringUtils.fromUtf8(payload))
```

Comment:
I wonder if this is too big for a log message, given that these payloads can be several MB